### ISSUE TITLE
More than 32k concurrent queries may cause errors

### ISSUE THREAD

**Issue by cvybhu** [scylladb/scylla-rust-driver#188]
If there happen to be more than 32 000 concurrent queries on a single connection we have no free stream ids for new queries and fail them with an error.
We could have an async [sempahore](https://docs.rs/tokio/1.2.0/tokio/sync/struct.Semaphore.html) with 32k permits to wait until a stream id is freed.
This could have a bad effect on performance, let's wait for some benchmarks to test different solutions.

**Comment by wprzytula** [comment]
Adding a semaphore would introduce nontrivial overhead. As no one ever complained about this in practice, I think we can close this. WDYT @Lorak-mmk?

**Comment by Lorak-mmk** [comment]
I don't think we should close this becasue this is a potential problem which may come up in the future.
We could implement this and benchmark the effect - but it will have a very low priority.

**Comment by Lorak-mmk** [comment]
Now that I think of it:
- 32k requests on a single connection is a pathological case that strongly signals there is a deeper issue
- Using a semaphore would just mask the problem, turning errors into increased client-side latencies

I think errors in this case are fine. Either user or retry policy can decide what to do with the query.
